core/drivers/crypto/caam/utils/utils_mem.c

```diff
@@ -0,0 +1,491 @@
+
+	return value;
+}
+
+#define MEM_TYPE_ZEROED BIT(0)
+#define MEM_TYPE_NORMAL BIT(1)
+#define MEM_TYPE_ALIGN BIT(2)
+
+#define MEM_TYPE_DEFAULT (MEM_TYPE_ZEROED | MEM_TYPE_NORMAL)
+
+#define CAAM_MEM_INFO
+#ifdef CAAM_MEM_INFO
+struct __packed mem_info {
+	void *addr;
+	size_t size;
+	uint8_t type;
+};
+
+#define MEM_INFO_SIZE ROUNDUP(sizeof(struct mem_info), sizeof(void *))
+#define OF_MEM_INFO(p) (struct mem_info *)((uint8_t *)(p) - MEM_INFO_SIZE)
+
+/*
+ * Allocate an area of given size in bytes. Add the memory allocator
+ * information in the newly allocated area.
+ *
+ * @size   Size in bytes to allocate
+ * @type   Type of area to allocate (refer to MEM_TYPE_*)
+ */
+static void *mem_alloc(size_t size, uint8_t type)
+{
+	struct mem_info *info = NULL;
+	vaddr_t ret_addr = 0;
+	void *ptr = NULL;
+	size_t alloc_size = 0;
+	uint32_t cacheline_size = 0;
+
+	MEM_TRACE("alloc %zu bytes of type %d", size, type);
+
+	/*
+	 * The mem_info header is added at just before the returned
+	 * buffer address
+	 *
+	 * --------------
+	 * |  mem_info  |
+	 * --------------
+	 * |  Buffer    |
+	 * --------------
+	 */
+	if (ADD_OVERFLOW(size, MEM_INFO_SIZE, &alloc_size))
+		return NULL;
+
+	if (type & MEM_TYPE_ALIGN) {
+		/*
+		 * In case of the buffer must be aligned on a cache
+		 * line, ensure that the size is aligned on a cache line
+		 *
+		 * Buffer address returned is moved up to a cache start offset,
+		 * hence add a cache line if the alloc_size is already aligned
+		 * otherwise the total buffer size will be too small
+		 */
+		cacheline_size = read_cacheline_size();
+		if (alloc_size == cacheline_size) {
+			if (ADD_OVERFLOW(alloc_size, cacheline_size,
+					 &alloc_size))
+				return NULL;
+		}
+		if (ADD_OVERFLOW(alloc_size,
+				 ROUNDUP(alloc_size, cacheline_size),
+				 &alloc_size))
+			return NULL;
+	}
+
+	ptr = malloc(alloc_size);
+	if (!ptr) {
+		MEM_TRACE("alloc Error - NULL");
+		return NULL;
+	}
+
+	if (type & MEM_TYPE_ZEROED)
+		memset(ptr, 0, alloc_size);
+
+	/* Calculate the return buffer address */
+	ret_addr = (vaddr_t)ptr + MEM_INFO_SIZE;
+	if (type & MEM_TYPE_ALIGN)
+		ret_addr = ROUNDUP(ret_addr, cacheline_size);
+
+	/*
+	 * Add the mem_info header
+	 */
+	info = OF_MEM_INFO(ret_addr);
+	info->addr = ptr;
+	info->size = size;
+	info->type = type;
+
+	MEM_TRACE("alloc returned %p -> %p", ptr, (void *)ret_addr);
+
+	return (void *)ret_addr;
+}
+
+/*
+ * Free allocated area
+ *
+ * @ptr  area to free
+ */
+static void mem_free(void *ptr)
+{
+	struct mem_info *info = NULL;
+
+	if (!ptr)
+		return;
+
+	info = OF_MEM_INFO(ptr);
+	MEM_TRACE("free %p info %p", ptr, info);
+	MEM_TRACE("free @%p - %zu bytes of type %d", info->addr, info->size,
+		  info->type);
+
+	free(info->addr);
+}
+#else
+/*
+ * Allocate an area of given size in bytes
+ *
+ * @size   Size in bytes to allocate
+ * @type   Type of area to allocate (refer to MEM_TYPE_*)
+ */
+static inline void *mem_alloc(size_t size, uint8_t type)
+{
```

Comment:
Sorry I don't see your point here. Which debug `mem_alloc()` are you talking about?

Reply:
Isn't `CAAM_MEM_INFO` enabling debug/info support in `mem_alloc()`/`mem_free()`?

Reply:
no. This is a feature replacing the removed functionality in the memory allocation library that was present in previous TEE version. This feature was allowing allocation of buffer with a start address aligned. In this case the returned address is not the address of the buffer allocated and so we have to keep the real buffer address to free it. My choice is to add a info header at the top of the buffer in which I keep the real buffer info.

Reply:
Do you need to keep these 2 implementations? If not, I strongly suggest to remove the unused one.

Reply:
for the moment yes except if you confirm that the memory allocator in TEE will never add back the function allocating a buffer with a start address aligned.

Reply:
As for current implementation, OP-TEE memory allocator (bget based) does not support `memalign()`. I cannot tell whether it will be supported or not in a near future. It depends on the need and contributions. Until then, you need specific alloc and free functions.

Reply:
ok, so I prefer to keep file as is.

Reply:
See #3116 for a discussion on aligned memory.